prusti-interface/src/specs/typed.rs

```diff
@@ -436,6 +438,23 @@ impl SpecGraph<ProcedureSpecification> {
         }
     }
 
+    pub fn add_purity<'tcx>(&mut self, purity: LocalDefId, env: &Environment<'tcx>) {
+        match self.get_constraint(purity, env) {
+            None => {
+                unreachable!(
+                    "separate purity defs are only used for type-conditional spec refinements"
+                )
+            }
+            Some(constraint) => {
+                let constrained_spec = self.get_constrained_spec_mut(constraint);
+                constrained_spec.kind =
+                    SpecificationItem::Inherent(ProcedureSpecificationKind::Pure);
+                // need to store this as well, since without pres or posts we couldn't find any def id with the trait bounds
```

Comment:
Is this the answer to the comment about the `purity` field? Is it because we still won't generate a spec item *just* for purity?

Reply:
It is indeed! From what I understand, the [current design](https://github.com/viperproject/prusti-dev/blob/a4ac54bbb747607ca9ab5d24f0013202a2e6e171/prusti-viper/src/encoder/mir/specifications/constraints.rs#L218-L221) (which we deemed out of scope to overhaul) looks at all the spec items, gathers `where` clauses as `ParamEnv`s, checks that there's only one such env (since we currently only support a single typecondspec refinement for any method), and returns it.

The problem is that to do so, it [runs through](https://github.com/viperproject/prusti-dev/blob/a4ac54bbb747607ca9ab5d24f0013202a2e6e171/prusti-viper/src/encoder/mir/specifications/constraints.rs#L234) all the pre- and postconditions listed in the `ProcedureSpecification`. I didn't touch this code initially until I realized it fails for ghost constraints that only establish purity and nothing else (no pres or posts) because the loop has nothing to run through. So [my solution](https://github.com/juliand665/prusti-dev/blob/0684867cc8765c8cc1abee796882393dd5761603/prusti-viper/src/encoder/mir/specifications/constraints.rs#L235) is to store and iterate over the purity spec item as well.